### PR TITLE
raftstore: only compact replicated log

### LINKED ISSUE #424: disable compact log when sending snapshot.
Let's consider following case:

First add two peers in raft region a, 1, 2, 1 is leader, then add 3.
1 may send snapshot to 3, but unlucky, after 3 finishes applying snapshot, 1 has already done a compact log, so 1 and 3 have gap logs, 1 will send snapshot  again, but if 1 did compact log before 3 applying snapshot, loop again.

So we should disable compacting log when doing snapshot. After leader guarantee that all peers catch up it, we can allow compacting again.

@ngaut @BusyJay

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @siddontang PTAL

close #424

### CODE REVIEW

**Comment by siddontang** [comment]
LGTM

**Comment by ngaut** [comment]
@xiang90 @siddontang PTAL

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -551,13 +552,22 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
             }
-
-            if applied_index - first_index <= self.cfg.raft_log_gc_threshold {
+            let replicated_idx = peer.raft_group
+                                     .status()
+                                     .progress
+                                     .values()
+                                     .map(|p| p.matched)
+                                     .min()
+                                     .unwrap();
```
Comment (tiancaiamao) [inline src/raftstore/store/store.rs]: hope the 'unwrap' here won't make the program crash.

Comment (BusyJay) [inline src/raftstore/store/store.rs]: No, it can't.

Comment (tiancaiamao) [inline src/raftstore/store/store.rs]: LGTM

**Comment by siddontang** [comment]
LGTM

**Comment by xiang90** [comment]
I am not sure if we really need to be this conservative to avoid snapshot loop.

Say we have three nodes in one cluster: A, B, and C. After C dies, A, B cannot compact its log due to this new rules. If C dies for a few seconds then it might be fine. But if C dies for minutes, then A and B might leave millions of log entries in memory or in the WAL files to be cleaned up. It might be OK or might not be OK. If it is OK for you, then probably it is the way to go here.

In my opinion, the purpose of this rule is to avoid snapshot sending loop when a snapshot can take long time to generate and send. I think a better approach is to only disable compaction when the leader finds out a active node (online one) needs a snapshot. If the snapshot is size limited then the generating and sending time is limited. Thus the time to disable compaction is limited since you can set up timeout properly.

I am not sure how you do snapshot right now. @siddontang mentioned another idea is to disable compaction when doing the snapshot. I assume you always need to do snapshot for checkpointing in some form, so that you can start compaction. Rr the node will not be able to restart. So I feel there is still a difference between disabling compaction when doing snapshot and when sending snapshot. But we might mean the same thing.

**Comment by xiang90** [comment]
BTW, I am not saying this solution does not work. It works as long as we are OK to have more number of entries. (probably you plan to gc long dead nodes... so it won't be a problem at all.)

Just provide some thing we might want to think about.

**Comment by siddontang** [comment]
Hi @xiang90 

We know that this solution can cause huge logs if the node is down for a long time, but we think it is ok now. 

I agree that we should find a better way to handle it later. Btw, what I mean `do` snapshot is for generate + send + apply, a whole flow, so we mean the same thing.

Maybe we can force compact log if we find the log is too huge and a node is down for a long time, and pd can find the node is dead and try to gc it automatically later.

**Comment by ngaut** [comment]
@xiang90 Thanks for your advice. It's complicate to handle all of the case. 
We still need to force compact log when log is huge.

@BusyJay Can we do that in this PR ?

**Comment by BusyJay** [comment]
Yes, but to be noted that adding a bound check will not solve issue #424.

**Comment by ngaut** [comment]
LGTM

**Comment by ngaut** [comment]
@xiang90 PTAL

**Comment by xiang90** [comment]
looks good. thanks.

**Pull request merged**